src/scenic/core/visibility.py

```diff
@@ -0,0 +1,503 @@
+                h_range = (math.pi+sphere_h_range[0], va_h_range[1])
+                view_ranges.append((h_range, overlapping_v_range))
+
+            if len(view_ranges) == 0:
+                return False
+
+        else:
+            # We can immediately check for impossible visbility/optimize which rays
+            # will be cast.
+
+            # Check if view range and spherical angles overlap in horizontal or
+            # vertical dimensions. If not, return False
+            if  (np.max(spherical_angles[:,0]) < -viewAngles[0]/2) or \
+                (np.min(spherical_angles[:,0]) >  viewAngles[0]/2) or \
+                (np.max(spherical_angles[:,1]) < -viewAngles[1]/2) or \
+                (np.min(spherical_angles[:,1]) >  viewAngles[1]/2):
+                return False
+
+            # Compute trimmed view angles
+            h_min = np.clip(np.min(spherical_angles[:,0]), -viewAngles[0]/2, viewAngles[0]/2)
+            h_max = np.clip(np.max(spherical_angles[:,0]), -viewAngles[0]/2, viewAngles[0]/2)
+            v_min = np.clip(np.min(spherical_angles[:,1]), -viewAngles[1]/2, viewAngles[1]/2)
+            v_max = np.clip(np.max(spherical_angles[:,1]), -viewAngles[1]/2, viewAngles[1]/2)
+
+            h_range = (h_min, h_max)
+            v_range = (v_min, v_max)
+
+            view_ranges = [(h_range, v_range)]
+
+        ## Generate candidate rays
+        candidate_ray_list = []
+
+        for h_range, v_range in view_ranges:
+            h_size = h_range[1] - h_range[0]
+            v_size = v_range[1] - v_range[0]
+
+            assert h_size > 0
+            assert v_size > 0
+
+            scaled_v_ray_count = math.ceil(v_size/(viewAngles[1]) * rayCount[1])
+            v_angles = np.linspace(v_range[0],v_range[1], scaled_v_ray_count)
+
+            # If altitudeScaling is true, we will scale the number of rays by the cosine of the altitude
+            # to get a uniform spread.
+            if altitudeScaling:
+                h_ray_counts = np.maximum(np.ceil(np.cos(v_angles) * h_size/(viewAngles[0]) * rayCount[0]), 1).astype(int)
+                h_angles_list = [np.linspace(h_range[0],h_range[1], h_ray_count) for h_ray_count in h_ray_counts]
+                angle_matrices = [np.column_stack([h_angles_list[i], np.repeat([v_angles[i]], len(h_angles_list[i]))]) for i in range(len(v_angles))]
+                angle_matrix = np.concatenate(angle_matrices, axis=0)
+            else:
+                scaled_h_ray_count = math.ceil(h_size/(viewAngles[0]) * rayCount[0])
+                h_angles = np.linspace(h_range[0],h_range[1], scaled_h_ray_count)
+                angle_matrix = np.column_stack([np.repeat(h_angles, len(v_angles)), np.tile(v_angles, len(h_angles))])
+
+            ray_vectors = np.zeros((len(angle_matrix[:,0]), 3))
+
+            ray_vectors[:,0] = -np.sin(angle_matrix[:,0])
+            ray_vectors[:,1] = np.cos(angle_matrix[:,0])
+            ray_vectors[:,2] = np.tan(angle_matrix[:,1])
```

Comment:
I was going to say that this is infinite at 90 degrees altitude, but it looks like numpy returns a very large but finite value, so everything works OK. Might be worth adding a comment about that to avoid confusion in the future.